강원대 BE_최현준 1단계 - 유효성 검사 및 에러 처리

Hyeonjun0527 opened the pull request:
죄송합니다..6시를 넘겨버렸습니다 ㅠㅠ..

바인딩 실패도 유효성 검사 실패로 판단하여 변경
오류를 4xx와 5xx로 구분하는게 에러의 원인을 인식하기에 명확함.
도메인 오류 요구사항 반영.
구체적인 에러 정보를 클라이언트에 제공하는 것에 의구심을 가졌음
그래서 사용자 친화적인 메세지를 전달하기로 함.
먼저 Exception 관련 코드가 변경되어 Exception을 Test하는 코드는 Deprecated하기로 함.
import문 정리
필요없는 주석 제거
통합 테스트 코드 작성

---
# 1차 리뷰
다음부턴 아래의 2차 리뷰 방식처럼 적겠습니다!

# 2차 리뷰

## 주요 변경 사항

2차에서는 기존의 mvc 아키텍처에서 헥사고날 아키텍처를 학습하고 적용해봄.
1차에서는 ProductService에게 Dto를 노출시키고 싶지 않아 중간계층을 만들어야겠다는 생각으로 
AddProductUseCase.java,DeleteProductUseCase.java,GetAllProductsUseCase.java
,GetProductUseCase.java,ProductApplicationService.java,UpdateProductUseCase.java를 만들었음.
이 UseCase에서는 Dto를 알고, 각 ProductService와 ProductMapper를 사용하는 상위 계층으로 둠으로서 
ProductService가 Dto를 모르게 하였음.
하지만 이때의 한계는 UseCase또한 Dto를 알게 됨. 단순 구현체이기 때문.

그래서 2차에서는 UseCase 인터페이스를 생성함.(헥사고날 아키텍처에서 port의 역할을 함).
그리고 다음과 같은 패키지 구조를 만듬
adapter
  |---------persistence
  |---------web
어댑터는 infrastructure 계층에 존재함.
어댑터는 내부 구현 기술에 대한 구현체를 뜻하며, jdbcClient와 같은 구체적인 기술과 HTTP 요청과 같은 구체적인 내용을 구현함.

ProductController가 헥사고날 아키텍처에서 Primary adapter.
ProductPersistenceAdapter는 Secondary adapter

Application Layer에 있는 구현체인 Use Case는 비지니스 흐름을 맡고
Domain Layer는 비즈니스 로직과 모델을 맡음.

아래와 같은 구조로 구현해보았다.

![image](https://github.com/user-attachments/assets/b7aa87d0-1adf-4a9c-8367-b7f56a72632e)


## 맞닥뜨린 문제 및 한계

현재 application은 내부기술을 몰라야 하는데 spring validation을 알고있음. 이에 대해서는 타협을 했다..또한 ProductInterator도 spring validation을 아는 dto를 사용중이다

## 중점적으로 리뷰 받고 싶은 곳
1.헥사고날 아키텍처를 적용한 부분
2.핸들러에서 메서드로 중복 분리한 부분

##추가 내용

Hyeonjun0527 committed:
chore(gift): REAME.md 작업

Hyeonjun0527 committed:
Feat(ProductTest.java): 테스트 코드 작성
시나리오1_하나의_상품_추가하고_조회하고_수정하고_삭제하기()
메서드 생성

Hyeonjun0527 committed:
Feat(ProductTest.java): 테스트 코드 작성
시나리오2_여러개의_상품_추가하고_하나_조회하고_삭제하고_추가하고_조회하기 작성

Hyeonjun0527 committed:
Refactor(ProductTest.java):
import문 정리
필요없는 주석 제거
(피드백반영)

Hyeonjun0527 committed:
Feat(ErrorCode):
바인딩 실패도 유효성 검사 실패로 판단하여 변경
오류를 4xx와 5xx로 구분하는게 에러의 원인을 인식하기에 명확함.
도메인 오류 요구사항 반영.
구체적인 에러 정보를 클라이언트에 제공하는 것에 의구심을 가졌음
그래서 사용자 친화적인 메세지를 전달하기로 함.

Hyeonjun0527 committed:
Feat(ErrorCode):
바인딩 실패도 유효성 검사 실패로 판단하여 변경
오류를 4xx와 5xx로 구분하는게 에러의 원인을 인식하기에 명확함.
도메인 오류 요구사항 반영.
구체적인 에러 정보를 클라이언트에 제공하는 것에 의구심을 가졌음
그래서 사용자 친화적인 메세지를 전달하기로 함.
먼저 Exception 관련 코드가 변경되어 Exception을 Test하는 코드는 Deprecated하기로 함.

jaenyeong commented on src/main/java/gift/common/exception/ErrorCode.java:
```diff
@@ -1,29 +1,66 @@
-    BINDING_FAILED("데이터 바인딩에 실패했습니다.");
+    
+    // 서버 내부 오류 (5xx)
+    UNEXPECTED_ERROR(HttpStatus.INTERNAL_SERVER_ERROR, "일시적인 오류가 발생했습니다. 잠시 후 다시 시도해주세요"),
+    INTERNAL_SERVER_ERROR(HttpStatus.INTERNAL_SERVER_ERROR, "서버에 문제가 발생했습니다. 지원팀에 문의해주세요"),
+    
+    // 클라이언트 요청 오류 (4xx)
+    RESOURCE_NOT_FOUND(HttpStatus.NOT_FOUND, "요청한 정보를 찾을 수 없습니다"),
+    INVALID_REQUEST_FORMAT(HttpStatus.BAD_REQUEST, "요청 형식이 올바르지 않습니다"),
+    MISSING_REQUIRED_PARAMETER(HttpStatus.BAD_REQUEST, "필수 정보가 누락되었습니다"),
+    
+    // 입력값 검증 오류 (4xx)
+    VALIDATION_FAILED(HttpStatus.BAD_REQUEST, "입력하신 정보를 다시 확인해주세요"),
+    INVALID_INPUT_VALUE(HttpStatus.BAD_REQUEST, "올바르지 않은 값이 입력되었습니다"),
+    REQUIRED_VALUE_MISSING(HttpStatus.BAD_REQUEST, "필수 입력값이 누락되었습니다"),
+
+    // 상품 도메인 검증 오류 (4xx)
+    PRODUCT_NAME_TOO_LONG(HttpStatus.BAD_REQUEST, "상품명은 15자 이하로 입력해주세요"),
+    PRODUCT_NAME_INVALID_CHARS(HttpStatus.BAD_REQUEST, "상품명에 사용할 수 없는 문자가 포함되어 있습니다"),
+    PRODUCT_NAME_KAKAO_RESTRICTED(HttpStatus.BAD_REQUEST, "카카오가 포함된 상품명은 사전 승인이 필요합니다"),
+    PRODUCT_PRICE_REQUIRED(HttpStatus.BAD_REQUEST, "상품 가격은 필수 입력값입니다"),
+    PRODUCT_PRICE_INVALID(HttpStatus.BAD_REQUEST, "상품 가격을 올바르게 입력해주세요"),
+    PRODUCT_NOT_FOUND(HttpStatus.NOT_FOUND, "상품 정보를 찾을 수 없습니다");
```
🟢 
점점 이런식으로 내용이 많아지면, http status 또는 에러 종류에 따라 enum 클래스를 나눠 선언하는 것도 도움이 될 수 있습니다 ㅎㅎ
물론 이런 말씀을 드리면서도 저였어도 이 정도면 굳이 나누진 않겠지만요 ㅎㅎㅎ

Hyeonjun0527 replied:
아 저는 이 enum을 종류에 따라 나눌 생각은 전혀 못해봤는데 리뷰해주신 덕분에 또 그런 선택지가 있다는 걸 알게 되네요.
한 번 적용해보면 학습에 도움이 될 거 같아 적용해보도록 하겠습니다.

jaenyeong commented on src/main/java/gift/common/exception/GlobalExceptionHandler.java:
```diff
@@ -3,55 +3,133 @@
+                    pd.setProperty("rejectedValue", fe.getRejectedValue());
+                    return pd;
+                })
+                .toList();
```
🟢 
중복이 있네요? ㅎㅎ 중복을 최소화 해주시면 좋겠습니다~

Hyeonjun0527 replied:
넵!

jaenyeong commented on src/main/java/gift/common/exception/GlobalExceptionHandler.java:
```diff
@@ -77,30 +155,42 @@ protected ResponseEntity<Object> handleHttpMessageNotReadable(
-                .property("path", req.getRequestURI())
-                .build();
+    //TODO : 현재 메세지 기반으로 에러코드를 추출하는데 해결책을 찾고 리팩토링이 필요함
+    private ErrorCode determineErrorCodeFromMessage(String message) {
```
🟢  
이 메서드는 ErrorCode에 있는게 OOP 측면에서 더 적합하지 않을까? 하는 생각이 드네요~
해당 위치로 이동하면 `fromMessage` 정도의 네이밍만 해도 되지 않을까 싶구요! 물론 이동한다면 코드는 조금 리팩터링 되겠죠?

결국 메시지로 적절한 에러코드를 매칭하여 반환하는 것이 해당 메서드의 책임이라고 볼 수 있는데 그래서 에러코드로 이동하는 건 어떻게 생각하시나요?

jaenyeong commented on src/main/java/gift/common/exception/GlobalExceptionHandler.java:
```diff
@@ -3,55 +3,133 @@
 
 @RestControllerAdvice
 @Order
 public class GlobalExceptionHandler extends ResponseEntityExceptionHandler {
```
👍 
전반적으로 공통으로 잘 모아서 구현해주신 것 같아요!

추가로 피드백은 아니고 일종의 의견?을 드려보면 RuntimeException handle은 따로 없어보이는데,
여기서 짚을 건 아니지만, 체크드 익셉션과 언체크드 익셉션은 트랜잭션을 적용했을때 롤백이 되냐 안되냐 차이도 있어서 보통 나눠 핸들링 하는 경우가 종종 있긴 합니다~

그리고 보통 이런 예외 핸들링이나 스프링 프로젝트의 전체 빈설정 등은 팀원 분들이 제각기 추가하게 되는 경우들이 많아요~
그래서 간단한 주석이라도 (무엇을 어떻게 핸들링하는 메서드인지?) 달아주시면 조금 더 좋을 것 같아요 ㅎㅎ

마지막으로 `@Order`를 달아두셔서 남기는 얘긴데, 많아지면 여러 기준을 가지고 예외 처리 핸들러를 구현하실 수도 있습니다
패키지, 또는 도메인이나 예외의 원인 등 공통적으로 묶어서 그룹핑 시킬 수 있는 경우에 나누면 더 보기 좋아지겠죠~
그때는 Order 애너테이션으로 순서를 지정할 수도 있겠네요 ㅎㅎ

(이 부분들은 수정 하시길 원하는 피드백은 아니에요!)

jaenyeong commented on src/main/java/gift/common/exception/GlobalExceptionHandler.java:
```diff
@@ -3,55 +3,133 @@
-                .property("code", ErrorCode.NULL_ERROR.name())
-                .property("field", ErrorField.NULL_POINTER.name())
+        ErrorCode errorCode = ErrorCode.REQUIRED_VALUE_MISSING;
+        return ErrorResponse.builder(ex, errorCode.getHttpStatus(), errorCode.getMessage())
+                .property("code", errorCode.name())
                 .property("path", req.getRequestURI())
                 .build();
```
🟢 
지금 보면 
```java
ErrorResponse.builder(ex, errorCode.getHttpStatus(), errorCode.getMessage())
                .property("code", errorCode.name())
                .property("path", req.getRequestURI())
                .build();
```
이 부분도 대부분 메서드에서 중복이라 줄여도 좋을 듯 합니다~

Hyeonjun0527 replied:
넵!

jaenyeong commented on src/test/java/gift/common/exception/GlobalExceptionHandlerTest.java:
```diff
@@ -19,6 +19,7 @@
 import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
 
 
+@Deprecated
```
❓ 
혹시 삭제를 하지 않고 해당 애너테이션을 태깅한 이유는 무엇일까요?
이걸 태깅하면서까지 테스트 코드를 남기는 이유가 있나요? 아시겠지만, 테스트 코드도 결국 관리가 필요한 코드입니다!
그래서 관리가 제대로 이루어지거나 어렵다 하시면 삭제하는 게 낫지 싶은데요!

저번 스텝에 코멘트를 남겨놓긴 했는데 해당 테스트 코드가 실패하고 있어서 이유가 궁금합니다~

Hyeonjun0527 replied:
어제 main의 에러처리 관련 코드를 고쳐서 테스트 코드가 작동하지 않았는데, 시간이 없었어서 테스트 코드를 지우기는 그렇고 다시 돌아가도록 수정하고 싶어서 저런 방식을 선택했습니다.. 이런 경우에 어떻게 해야할 지를 몰라 저 어노테이션을 태깅했습니다 더 좋은 방식은 떠오르는거로 stash공간에 넣고 제외하고 push하는 방법이 있을 거 같은데 이런 경우에 멘토님은 어떤 방식을 사용하시나요?

jaenyeong commented on src/test/java/gift/domain/product/integration/ProductTest.java:
```diff
@@ -0,0 +1,145 @@
+
+        for (int i = 1000; i <= 5000; i+=1000) {
+            productRequests.add(new ProductRequest(NAME, i, IMAGE_URL));
+        }
```
🔴 
혹시 이 부분이 왜 BeforeEach로 필요할까요?
제가 봤을 때 첫번째 테스트 메서드에서는 5개나 미리 세팅할 필요가 없다라고 생각 되어서요~

그래서 두번째 메서드에서만 필요하다면 해당 테스트 메서드 내부에서 세팅하는 것이 더 낫다고 생각되네요!
이유는 @BeforeEach는 결국 매번 테스트 메서드들이 실행될 때 마다 실행될건데 불가피하게 이런 작업들을 매번 실행한다면, 테스트 실행 성능에 영향을 끼치게 되어서요! 지금은 생성하는 객체의 수도 적고, 컴퓨팅 리소스가 많이 필요한 작업을 하지는 않지만 이런 테스트코드들이 모이고 모이면 결국은 성능에 영향을 끼치게 됩니다 ㅠㅠ

아무쪼록 request를 각 테스트 메서드에서 필요한 만큼 세팅하여 사용한다면 클래스 멤버변수로 `productRequests` 를 갖고 있을 필욘 없겠죠~

Hyeonjun0527 replied:
BeforeEach에서 객체를 미리 생성하지 않게 하여 각 테스트 메서드가 결합하지 않도록 분리하였습니다!

jaenyeong commented on src/test/java/gift/domain/product/integration/ProductTest.java:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    @Test
+    void 시나리오1_하나의_상품_추가하고_조회하고_수정하고_삭제하기() {
```
🔴 
지금 모든 테스트 메서드에서 너무 많은 케이스를 검증하고 있습니다~
이를 분리하여 각각의 테스트 메서드를 따로 만들어 주시면 좋겠습니다 ㅎㅎ

추가로 AAA, BDD 와 같은 테스트 작성 패턴이 있는데요~ 이것도 함께 참고하셔서 각각의 테스트 코드들을 작성해 보시는 것도 추천 드립니다~

Hyeonjun0527 replied:
BDD 테스트 작성 패턴으로 리팩토링을 진행해보았습니다!!!

jaenyeong reviewed (changes requested):
역시나 너무 잘 해주셨고, 많은 고민을 하셨다는 게 코드에서 들어나네요 ㅎㅎ

프로덕션 코드에서는 반드시 수정해야한다 라고 할만한 점은 딱히 없었습니다~
🟢  코멘트는 꼭 반영할 필요는 없습니다 ㅎㅎ
다만 테스트 코드를 작성하셨기에 이쪽으로 수정 사항을 조금 남겨봤습니다!

Hyeonjun0527 committed:
refactor(product): 유스케이스 계층 도입 및 아키텍처 개선

Product 관련 기능 유지보수성, 테스트 용이성, 코드 명확성 향상을 위해 대규모 리팩토링 진행

주요 변경 사항:

아키텍처 변경:

컨트롤러의 책임을 URL 매핑과 입력 검증으로 한정
서비스는 오로지 도메인 로직에만 집중(서비스와 Request DTO와의 결합 제거)
개별 유스케이스들을 총괄하는 ProductApplicationService 도입

컨트롤러는 이제 ProductApplicationService에만 의존하여 lean해짐

도메인 모델과 DTO 간 변환을 담당하는 ProductMapper를 빈(Bean)으로 전환하여 재사용성 향상

DTO 및 도메인 모델:

불변성과 간결성을 위해 ProductRequest, ProductResponse를 클래스에서 record 타입으로 마이그레이션

Product 모델 생성 로직을 캡슐화하기 위해 팩토리 메서드(of) 패턴 적용

테스트:

Deprecated로 사라질 예정인 MockBean을 제거.

통합 테스트(ProductTest)에 BDD(Given-When-Then) 스타일과 AAA 패턴 적용하여 테스트 목적과 시나리오 가독성 향상

Hyeonjun0527 committed:
refactor(validation): 메시지 키워드 추가 및 에러코드 매핑 로직 개선

- ValidationMessages 클래스를 추가하여 검증 메시지 키워드를 중앙화
- ProductRequest의 검증 메시지를 ValidationMessages 사용으로 변경
- ErrorCode에 키워드 기반 매칭 로직 추가 및 개선
  - 기존 하드코딩된 에러 매핑 로직 제거
  - 메시지 키워드를 활용한 매칭 방식으로 대체
- GlobalExceptionHandler에서 메시지 기반 에러코드 결정 로직을 ErrorCode의 새 메서드로 교체

Hyeonjun0527 committed:
refactor(product): 상품 페이징 처리 및 UI 개선

- ProductController 수정: pageable을 RequestParam에서 Spring Pageable로 변경
- Pageable 클래스에 기본 생성자 추가 및 메서드 추가 (page, size 설정 가능하도록 변경)
- 프론트엔드 JS 리팩터링:
  - 반복 요소 제거: ui 객체로 DOM 요소 관리
  - 함수 분리 (setupAddModal, setupEditModal)
  - 에러 처리 추가 및 공통화: handleApiError 함수 추가
- 코드 가독성 및 중복 제거를 위한 리팩터링 작업 진행함

Closes #31

jaenyeong commented on src/main/java/gift/application/product/AddProductUseCase.java:
```diff
@@ -0,0 +1,28 @@
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+public class AddProductUseCase {
```
👍 
UseCase와 Mapper를 보니 헥사고날 아키텍처와 같은 형태로 구성하신 것 같네요!
사실 지금 과제 수준에는 조금 과한 설계긴 한데 ㅎㅎㅎ 학습 중이시니 다양한 형태로 구성해 보시는 것도 실력 향상과 인사이트에 도움이 될 듯 합니다 ㅎㅎ

다만 조금 이상한 부분을 말씀드리면 
과하게 계층을 나눴다 뭐 이런 얘기는 지금 잠시 접어두고, `ProductApplicaitonService`에서 각 유스케이스 클래스들을 참조하고, 또 이 클래스들이 프로덕트 서비스 클래스를 참조하는 플로우인데요! 즉 의존성이 `Service` -> `UseCase` -> `Service` 라는 얘기인데요 많이 이상하죠? 서비스, 리포지터리는 스프링 프레임워크에서 구성한 레이어드 아키텍처 요소이고, 유스케이스는 직접 만든 계층(레이어)인데 혼종이 되어 버렸죠..? 각 계층에 대한 레이어드를 조금 더 고민하셔서 바꿔보시길 바라겠습니다 ㅎㅎ 추가로 `@Service` 애너테이션을 볼 때 UseCase 클래스들 또한 해당 아키텍처를 모르는 사람이 봤을땐 서비스레이어라고 오해할 수 있을 것 같아요! 그래서 같은 레이어 끼리는 가급적 참조를 안하는 형태로 계층을 추가해 구성한다면 `UseCase` 전용 애너테이션을 만들어서 사용할 수도 있겠네요~

jaenyeong commented on src/main/java/gift/application/product/ProductApplicationService.java:
```diff
@@ -0,0 +1,55 @@
+    @Transactional
+    public void deleteProduct(Long id) {
+        deleteProductUseCase.execute(id);
+    }
```
🔴 
트랜잭션을 붙여 주신 건 너무 좋습니다~ 원래 조금 더 뒤에 커리큘럼에 있어서 그때 말씀드릴려고 했던 부분이었는데요!

다만 지금 해당 클래스에서 `@Transactional` 애너테이션이 붙어 있는데요! 문제는 각 계층마다 다 붙어 있어서 원하시는 의도를 모르겠습니다~
스프링의 트랜잭션에 대해 공부해 보시면 아시겠지만 전파 옵션을 생성하여 새 트랜잭션을 여는 것도 아니고, 가장 최초에 생성되는 트랜잭션 하나로 묶일 것이라 지금은 각 레이어에 다 붙이는게 의미가 없습니다!
그래서 현준님이 트랜잭션이 적용되어야 할 가장 적합한 레이어라고 생각하는 곳에만 태깅해 주시면 좋겠습니다~

jaenyeong commented on src/main/java/gift/common/exception/GlobalExceptionHandler.java:
```diff
@@ -3,55 +3,133 @@
+                    pd.setProperty("field", fe.getField());
+                    pd.setProperty("rejectedValue", fe.getRejectedValue());
+                    return pd;
+                })
```
🔴 
여전히 중복이 남아 있네요 ㅠㅠ

Hyeonjun0527 replied:
중복 제거하였습니다!

jaenyeong commented on src/main/java/gift/domain/product/mapper/ProductMapper.java:
```diff
@@ -0,0 +1,34 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class ProductMapper {
```
🟢 
지금 이 구조에서 `ProductMapper` 는 별도 인스턴스가 필요 없어 보여서, 각 메서드들도 static 메서드로 구현하여 사용하셔도 될 것 같네요!

Hyeonjun0527 replied:
이 부분은 꼭 다음에 추후 적용하겠습니다 ㅠ

jaenyeong commented on src/main/java/gift/common/validation/ValidationMessages.java:
```diff
@@ -0,0 +1,19 @@
+package gift.common.validation;
+
+public final class ValidationMessages {
```
🟢 
단순히 메세지 문자열만 보관한다면 꼭 코드가 아니라 해당 값들을 스프링의 message source 를 활용하실 수도 있습니다 ㅎㅎ
또한 해당 클래스는 validation을 위한 메시지지만, 메시지 중에서 유효성 검사 부분에서 참조하는 "메시지"기 때문에 패키지명으로 봤을땐 `validation` 보다는 `message` 같은 패키지명이 더 적합하지 않을까라는 생각이 드네요~

Hyeonjun0527 replied:
넵 조언 감사드립니다!!

jaenyeong commented on src/main/java/gift/common/pagination/Pageable.java:
```diff
@@ -24,6 +29,20 @@ public int getSize() {
         return size;
     }
 
+    public void setSize(int size) {
+        if (size < 1) {
+            throw new IllegalArgumentException("페이지 크기는 1보다 작을 수 없다.");
+        }
+        this.size = size;
+    }
+
+    public void setPage(int page) {
+        if (page < 0) {
+            throw new IllegalArgumentException("페이지 번호는 0보다 작을 수 없다.");
+        }
+        this.page = page;
+    }
+
```
🟢 
지금 두 세터는 단순히 값을 세팅만 하지 않고, 유효성 검사를 통해 예외를 반환하기도 하는데요~
이런 경우 `setPageOrThrow` 처럼 예외가 던져질 수 있음을 명시해 주시거나, `setCorrectPageValue` 등 별도의 로직이 있다는 것을 명시해 주시면 이를 활용하는 다른 개발자분들에게 도움이 될 것 같아요 ㅎㅎ

Hyeonjun0527 replied:
넵 조언 감사드립니다!! 이부분도 인지하겠습니다

jaenyeong reviewed (changes requested):
리뷰가 조금 늦었습니다 ㅠ
🔴 뱃지 코멘트 한 번 더 확인 부탁드릴게요~
나머지도 코멘트들은 여유 되시면 반영을 고민해 봐주시구요!

> 어제 main의 에러처리 관련 코드를 고쳐서 테스트 코드가 작동하지 않았는데, 시간이 없었어서 테스트 코드를 지우기는 그렇고 다시 돌아가도록 수정하고 싶어서 저런 방식을 선택했습니다.. 이런 경우에 어떻게 해야할 지를 몰라 저 어노테이션을 태깅했습니다 더 좋은 방식은 떠오르는거로 stash공간에 넣고 제외하고 push하는 방법이 있을 거 같은데 이런 경우에 멘토님은 어떤 방식을 사용하시나요?

여러 방법이 있겠지만, 가장 간단하게는 주석에 해당 내용/히스토리를 남겨두실 수도 있을 듯하고용
저라면 그냥 삭제할 것 같아요~ 왜냐하면 실무에서는 테스트 커버리지를 측정하거나, 코드 프리 푸시 단계에서 빌드를 다 돌려서 통과가 되지 못하면 푸시를 못하게 블락하는 훅을 설정하는 경우도 있어서요~
꼭 이런거에 걸리지 않더라도 개인적으로 좀 컴팩션하고 깔끔한 것을 추구하는 편이라서요 ㅎㅎ

참고로 깃에 이런 것들이 다 이력에 남으니, 삭제하고 나중에 여유될 때 로그 보고 복구할 수 있습니다~

Hyeonjun0527 committed:
refactor(product): 헥사고날 아키텍처 적용 및 기존 모듈 제거

- 헥사고날 아키텍처
spring-gift-wishlist/
└── src/
    └── main/
        └── java/
            └── gift/
                ├── common/                (공통 코드: 예외, 어노테이션, 유틸 등)
                └── product/
                    ├── adapter/           (인프라 계층: 외부와 연결)
                    │   ├── web/           (Primary Adapter: Controller 등)
                    │   └── persistence/   (Secondary Adapter: Repository 구현 등)
                    │
                    ├── application/       (애플리케이션 계층: 비즈니스 워크플로우)
                    │   ├── port/          (포트: 계층 간 추상화 인터페이스)
                    │   │   ├── in/        (Input Port: Use Case 인터페이스)
                    │   │   └── out/       (Output Port: 외부 의존 인터페이스)
                    │   └── usecase/       (Use Case 구현체: 실제 워크플로우/조정)
                    │
                    └── domain/            (도메인 계층: 핵심 비즈니스 로직, 엔티티, 도메인 서비스)

- 기존 애플리케이션 서비스 계층 삭제
- 포트와 어댑터 기반 아키텍처로 리팩터링: AddProductUseCase, DeleteProductUseCase 등 분리
- @Adapter와 UseCase 어노테이션 추가로 어댑터 역할 명시
- SOLID의 ISP 원칙에 따라 컨트롤러가 각 유스케이스 인터페이스를 개별 주입받도록 변경

Hyeonjun0527 committed:
docs(README): 리드미 반영

jaenyeong reviewed (approved):
고생하셨습니다~
헥사고날이 많이 복잡할텐데도 끈기 있게 해주셨네요 ㅎㅎ
개인적으로는 Mapper 위치 정도만 조금 고민 포인트이긴 하네요 유스케이스 패키지 하위에 있는게 조금 어색한 것 같긴 해서요~

MD 파일에 정리도 너무 잘해주셨네요 ㅎㅎ
다음 스텝으로 넘어갈게요~
